src/validations/html/rules/MismatchedTag.js

```diff
@@ -0,0 +1,139 @@
+import {List, Map, Record, Set} from 'immutable';
```

Comment:
@tenyoung795 OK, that’s a good point about custom equality. As far as I can tell you‘re correct that the native data structures can only work on exact object/primitive value equality.

That said, a couple of further thoughts on this:

* The fact that `Immutable` collections do the right thing here is fairly tangential to the core competency of `Immutable`, i.e. being immutable! And my previous observation stands that actually dealing with immutability in this module does make some of the code harder to follow. How about using e.g. [collections.js](https://github.com/montagejs/collections), which provides normal mutable collections but does have the ability to define custom equality for keys?
* If we don’t want to introduce a new library to solve this problem, I think we should be consistent about the type of collections we use—meaning I would advocate changing `_openTagStack` back to an `Immutable.List` to be consistent with everything else.